compiler/noirc_frontend/src/monomorphization/debug_types.rs

```diff
@@ -0,0 +1,70 @@
+use crate::hir_def::types::Type;
+pub use noirc_errors::debug_info::{Types, VariableTypes, Variables};
+use noirc_printable_type::PrintableType;
+use std::collections::HashMap;
+
+/// We keep a collection of the debug variables and their types in this
+/// structure. The fe_var_id refers to the ID given when inserting the
+/// instrumentation probe. This variable does not have a type yet and hence it
+/// can be instantiated multiple types if it's in the context of a generic
+/// variable. The var_id refers to the ID of the instantiated variable which
+/// will have a valid type.
+#[derive(Debug, Clone, Default)]
+pub struct DebugTypes {
+    fe_to_vars: HashMap<u32, u32>,          // fe_var_id => var_id
+    variables: HashMap<u32, (String, u32)>, // var_id => (var_name, type_id)
+    types: HashMap<PrintableType, u32>,
+    id_to_type: HashMap<u32, PrintableType>,
```

Comment:
I'm not sure. It is just odd seeing `PrintableType`s used for purposes outside of printing. It may be something to mark to change / re-examine for later. `hir_def::Type` should always be fully resolved since it is the result of resolving an `UnresolvedType`. If you meant any type variables should be followed, there is `.follow_bindings()` to follow all Bound type variable links if that helps. We should also be able to make types serializable as well but I'll leave the ultimate decision to you here.